Move private forward grad mode helpers to torch.autograd.forward_ad

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90418
* #90415
* #90077
* __->__ #90240
* #90037

Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

zou3519 committed:
Move private forward grad mode helpers to torch.autograd.forward_ad

Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

soulitzer commented on torch/autograd/forward_ad.py:
```diff
@@ -181,3 +182,18 @@ def __enter__(self):
+# Private helper function to enable or disable fwd grad.
+# If you're a user and want to use this, please file an issue to discuss the use case.
+@contextlib.contextmanager
+def _enable_fwd_grad(enabled=True):
```
Maybe we should match the API of the current `_set_grad_enabled`, i.e. name it `_set_fwd_grad_enabled` and don't have a default value. (The `enable_grad` API currently doesn't take any arguments and always sets it to True.)

zou3519 replied:
Sure

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 committed:
Update on "Move private forward grad mode helpers to torch.autograd.forward_ad"


Motivation
- These were previously defined in functorch. They are not
functorch-specific, so I'm moving them to torch.autograd.forward_ad and
the autograd python bindings.
- I need this to avoid some of my cyclic import problems.

Should these be public APIs? Probably. Though this needs discussion, so
punting it to the future.

Test Plan:
- moved the tests of these from test/functorch/test_eager_transforms.py
to test/test_autograd.py

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "test failures look unrelated"

albanD commented:
Comments like https://github.com/pytorch/pytorch/blob/5c133c5744ed1ce3c471cc99c1f140282a3fb1bc/torch/csrc/autograd/autograd_meta.cpp#L275 need to be updated following this

zou3519 commented:
@albanD will do, added to followups (https://github.com/pytorch/pytorch/issues/90224)
